src/System.CommandLine.Hosting/HostingExtensions.cs

```diff
@@ -36,15 +36,17 @@ public static class HostingExtensions
                 hostBuilder.UseConsoleLifetime();
                 configureHost?.Invoke(hostBuilder);
 
+                var invokeCancel = invocation.GetCancellationToken();
```

Comment:
What's the difference between calling this, and calling `invocation.AddCancellationHandling()`?

Reply:
@tmds AddCancellationHandling() can only be called once. And it's called that one time in the InvokeContext constructor. It is also internal, and thus not callable from other assemblies.

Reply:
calling it multiple times will throw an `InvalidOperationException`

Reply:
> calling it multiple times will throw an InvalidOperationException

I know. I put the exception there.

The idea is: this is called once when the invocation context becomes cancellable, from then on-wards the user needs to use that CancellationToken.

>  And it's called that one time in the InvokeContext constructor.

It doesn't call `AddCancellationHandling`, but adds a factory method for when someone binds to a `CancellationToken`.
You can call `AddCancellationHandling` in `UseHost`, it shouldn't throw.

Reply:
@tmds yeah, but adding Cancellation Handling is not a concern that `UseHost` should deal with. You'd call `CancelOnProcessTermination` to add that capability.

Reply:
more importantly, `UseHost` would otherwise exclude usage for `CancelOnProcessTermination` or **ANY** other middlware that wanted to implement their own cancellation handling.

There is simply no need for `UseHost` to that, when it can simply get the CancellationToken (however that is configured) from the `InvocationContext`. `UseHost` wants to use `CancellationToken` it does not need to actually set it up (why would it need to?).

Reply:
My understanding is the method added here is the same thing as:
```
public CancellationToken GetCancellationToken()
  => AddCancellationHandling();
```
so I don't understand what you mean by:

> more importantly, UseHost would otherwise exclude usage for CancelOnProcessTermination or ANY other middlware that wanted to implement their own cancellation handling.

> it does not need to actually set it up (why would it need to?).

The pattern is to distinguish between invocations that can be cancelled and those that do not. In the latter case we don't want `CancelOnProcessTermination` to ignore SIGINT/SIGTERM.

Reply:
@tmds okay, yes I can see what you mean. However, you could still override the `CancellationToken` in the `BindingContext` service provider by calling `AddService<CancellationToken>()`. Not sure anyone would want to do that, but technically possible.

More importantly: You still cannot simply call `AddCancellationHandling`. The method is internal, an thus not accessible.

`GetCancellationToken` mimics what the BindingContext would actually do when it needed to fill in a `CancellationToken` value, so that is the more sensible way to do it (for a public method). Since `AddCancellationHandling` is internal, it is a mere implementation detail and cannot guarantee a correct value, whereas `BindingContext` actually IS the source of truth here.

Reply:
> However, you could still override the CancellationToken in the BindingContext service provider by calling AddService<CancellationToken>(). Not sure anyone would want to do that, but technically possible.

This isn't a feature.

> More importantly: You still cannot simply call AddCancellationHandling. The method is internal, and thus not accessible.

We can make it public.

> GetCancellationToken mimics what the BindingContext would actually do when it needed to fill in a CancellationToken value, so that is the more sensible way to do it (for a public method). Since AddCancellationHandling is internal, it is a mere implementation detail and cannot guarantee a correct value, whereas BindingContext actually IS the source of truth here.

The ServiceProvider is for implementing the typed argument binding. I don't see the value of using it here, and it makes a complex `GetCancellationToken` implementation.